src/data/real/sqrt.lean

```diff
@@ -97,15 +97,98 @@ end nnreal
 namespace real
 
 /-- An auxiliary sequence of rational numbers that converges to `real.sqrt (mk f)`.
+This is Heron's method for computing square roots.
 Currently this sequence is not used in `mathlib`.  -/
 def sqrt_aux (f : cau_seq ℚ abs) : ℕ → ℚ
-| 0       := rat.mk_nat (f 0).num.to_nat.sqrt (f 0).denom.sqrt
-| (n + 1) := let s := sqrt_aux n in max 0 $ (s + f (n+1) / s) / 2
+| 0       := 1
+| (n + 1) := let s := sqrt_aux n in (s + (max 0 (f (n+1))) / s) / 2
 
-theorem sqrt_aux_nonneg (f : cau_seq ℚ abs) : ∀ i : ℕ, 0 ≤ sqrt_aux f i
-| 0       := by rw [sqrt_aux, rat.mk_nat_eq, rat.mk_eq_div];
-  apply div_nonneg; exact int.cast_nonneg.2 (int.of_nat_nonneg _)
-| (n + 1) := le_max_left _ _
+theorem sqrt_aux_nonneg (f : cau_seq ℚ abs) (i : ℕ): 0 ≤ sqrt_aux f i :=
+begin
+  induction i with i hyp,
+  { unfold sqrt_aux, exact zero_le_one},
+  { unfold sqrt_aux,
+    simp, cancel_denoms,
+    have t : 0 ≤ max 0 (f (i + 1)) / sqrt_aux f i := div_nonneg (le_max_left 0 _) hyp,
+    exact add_nonneg hyp t, },
+end
+
+theorem sqrt_aux_ne_zero_step (f : cau_seq ℚ abs) (i : ℕ) (h : sqrt_aux f i ≠ 0) :
+  sqrt_aux f (i + 1) ≠ 0 :=
+begin
+  unfold sqrt_aux,
+  simp only [ne.def, div_eq_zero_iff, bit0_eq_zero, one_ne_zero, or_false],
+  apply ne_of_gt,
+  have t : 0 ≤ (max 0 (f (i + 1))) / sqrt_aux f i,
+  { rw div_nonneg_iff,
+    left,
+    exact ⟨le_max_left 0 _, sqrt_aux_nonneg f i⟩, },
+  have u : 0 < sqrt_aux f i := (ne.symm h).le_iff_lt.mp (sqrt_aux_nonneg f i),
+  exact lt_add_of_pos_of_le u t
+end
+
+@[simp]
+theorem sqrt_aux_eq_zero_iff_step (f : cau_seq ℚ abs) (i : ℕ) :
+  sqrt_aux f (i + 1) = 0 ↔ sqrt_aux f i = 0 :=
+begin
+  split,
+  { intros hyp,
+    { by_contradiction,
+      exact sqrt_aux_ne_zero_step f i h hyp, }, },
+  { intro prev_zero,
+    unfold sqrt_aux, rw [prev_zero], simp, },
+end
+
+@[simp]
+theorem sqrt_aux_eq_zero_iff (f : cau_seq ℚ abs) (i : ℕ) : sqrt_aux f i ≠ 0 :=
+begin
+  induction i with i hyp,
+  { unfold sqrt_aux,
+    norm_num, },
+  { simp [hyp], },
+end
+
+/-- A simplified version of Cauchy-Schwarz on a two-element set, which we prove separately to
+reduce the import graph. -/
+theorem rat.cauchy_schwarz (a b : ℚ) : 4 * a * b ≤ (a + b) ^ 2 :=
+begin
+  suffices pos: 0 ≤ (a + b) ^ 2 - 4 * a * b, exact sub_nonneg.mp pos,
+  calc 0 ≤ (a - b) ^ 2 : sq_nonneg (a - b)
+    ... = (a + b) ^ 2 - 4 * a * b : by ring,
+end
+
+lemma sq_div_self {a : ℚ} (a_nonzero : a ≠ 0) : a ^ 2 / a = a :=
+by rw [pow_two, div_eq_iff a_nonzero]
+
+theorem sqrt_aux_overestimate (f : cau_seq ℚ abs) {i : ℕ} (i_pos : 0 < i) :
+  f i ≤ (sqrt_aux f i) ^ 2 :=
+begin
+  cases i,
+  { linarith, },
+  { unfold sqrt_aux,
+    simp,
+    norm_num,
+    cases le_or_gt (f (i + 1)) 0,
+    { rw max_eq_left h,
+      simp,
+      calc f (i + 1) ≤ 0 : h
+        ... ≤ sqrt_aux f i ^ 2 / 4 : div_nonneg (sq_nonneg _) (by norm_num), },
+    { rw max_eq_right_of_lt h,
+      by_cases sqrt_aux f i = 0,
+      { simp at h, exfalso, exact h, },
```

Comment:
Using `by_cases` and then immediately proving one case can't happen is not the most readable: you can do
```
have h : sqrt_aux f i ≠ 0,
{ ... },
```